docker/build_artifacts/sagemaker/python_service.py

```diff
@@ -10,54 +10,236 @@
 # language governing permissions and limitations under the License.
-import signal
-from contextlib import contextmanager
-
+import bisect
 import importlib.util
 import json
 import logging
 import os
-import re
-from collections import namedtuple
-from time import sleep
+import subprocess
+import time
 
 import falcon
 import requests
-from proxy_client import GRPCProxyClient
 
-from multi_model_utils import MultiModelException
+from multi_model_utils import lock, timeout, MultiModelException
+import tfs_utils
 
-INFERENCE_SCRIPT_PATH = '/opt/ml/model/code/inference.py'
+SAGEMAKER_MULTI_MODEL_ENABLED = os.environ.get('SAGEMAKER_MULTI_MODEL', 'false').lower() == 'true'
+INFERENCE_SCRIPT_PATH = '/opt/ml/{}/code/inference.py'.format('models'
+                                                              if SAGEMAKER_MULTI_MODEL_ENABLED
+                                                              else 'model')
+PYTHON_PROCESSING_ENABLED = os.path.exists(INFERENCE_SCRIPT_PATH)
+SAGEMAKER_BATCHING_ENABLED = os.environ.get('SAGEMAKER_TFS_ENABLE_BATCHING', 'false').lower()
 MODEL_CONFIG_FILE_PATH = '/sagemaker/model-config.cfg'
 TFS_GRPC_PORT = os.environ.get('TFS_GRPC_PORT')
 TFS_REST_PORT = os.environ.get('TFS_REST_PORT')
+SAGEMAKER_TFS_PORT_RANGE = os.environ.get('SAGEMAKER_SAFE_PORT_RANGE')
+
 
 logging.basicConfig(level=logging.INFO)
 log = logging.getLogger(__name__)
 
-DEFAULT_CONTENT_TYPE = 'application/json'
-DEFAULT_ACCEPT_HEADER = 'application/json'
 CUSTOM_ATTRIBUTES_HEADER = 'X-Amzn-SageMaker-Custom-Attributes'
 
-Context = namedtuple('Context',
-                     'model_name, model_version, method, rest_uri, grpc_port, '
-                     'custom_attributes, request_content_type, accept_header, content_length')
+
+def default_handler(data, context):
+    """A default inference request handler that directly send post request to TFS rest port with
+    un-processed data and return un-processed response
+
+    :param data: input data
+    :param context: context instance that contains tfs_rest_uri
+    :return: inference response from TFS model server
+    """
+    response = requests.post(context.rest_uri, data=data)
+    return response.content, context.accept_header
 
 
-class InvocationResource(object):
+class PythonServiceResource(object):
 
     def __init__(self):
-        self._tfs_default_model_name = os.environ['TFS_DEFAULT_MODEL_NAME']
-        self._tfs_grpc_port = TFS_GRPC_PORT
-        self._tfs_rest_port = TFS_REST_PORT
+        if SAGEMAKER_MULTI_MODEL_ENABLED:
+            self._model_tfs_rest_port = {}
+            self._model_tfs_grpc_port = {}
+            self._model_tfs_pid = {}
+            self._tfs_ports = self._parse_sagemaker_port_range(SAGEMAKER_TFS_PORT_RANGE)
+        else:
+            self._tfs_grpc_port = TFS_GRPC_PORT
+            self._tfs_rest_port = TFS_REST_PORT
+
+        self._tfs_enable_batching = SAGEMAKER_BATCHING_ENABLED == 'true'
```

Comment:
nit: use double-quotes instead of single quotes throughout? (use same formatting as PySDK does?)

Reply:
This is a legacy problem that this packages uses single quotes from beginning. I also think we should follow Python SDK's convention to use double quotes. I will open a separate PR to address this issue, as well as the inconsistent usage of "utf8" and ``encodings.utf_8.getregentry().name``